cmd/lint-panics/analyzer/analyzer.go

```diff
@@ -0,0 +1,247 @@
+		)
+
+		defPos, err := p.checkGoroutineDefinition(ctx, pos, pass)
+		if err != nil {
+			p.logLinterError(pass, defPos, fun.Sel.Pos(), err)
+		}
+
+	case *ast.Ident: // function call
+		pos := pass.Fset.Position(fun.Pos())
+		p.logger.Info("found function call as goroutine",
+			zap.String("functionName", fun.Name),
+			utils.ZapURI(pos.Filename, pos.Line),
+			zap.Int("column", pos.Column),
+		)
+
+		defPos, err := p.checkGoroutineDefinition(ctx, pos, pass)
+		if err != nil {
+			p.logLinterError(pass, defPos, fun.Pos(), err)
+		}
+
+	default:
+		p.logger.Error("unexpected goroutine type",
+			zap.String("type", fmt.Sprintf("%T", fun)),
+		)
+	}
+}
+
+func (p *Analyzer) parseFile(path string, pass *analysis.Pass) (*ast.File, error) {
+	logger := p.logger.With(zap.String("path", path))
+
+	src, err := os.ReadFile(path)
+	if err != nil {
+		logger.Error("failed to open file", zap.Error(err))
+	}
+
+	file, err := goparser.ParseFile(pass.Fset, path, src, 0)
+	if err != nil {
+		logger.Error("failed to parse file", zap.Error(err))
+		return nil, err
+	}
+
+	return file, nil
+}
+
+func (p *Analyzer) checkGoroutine(body *ast.BlockStmt) error {
+	if body == nil {
+		p.logger.Warn("missing function body")
+		return nil
+	}
+	if len(body.List) == 0 {
+		// empty goroutine is weird, but it never panics, so not a linter error
+		return nil
+	}
+
+	deferStatement, ok := body.List[0].(*ast.DeferStmt)
+	if !ok {
+		return errors.New("first statement is not defer")
+	}
+
+	selectorExpr, ok := deferStatement.Call.Fun.(*ast.SelectorExpr)
+	if !ok {
+		return errors.New("first statement call is not a selector")
+	}
+
+	firstLineFunName := selectorExpr.Sel.Name
+	if firstLineFunName != Pattern {
+		return errors.Errorf("first statement is not %s", Pattern)
+	}
+
+	return nil
+}
+
+func (p *Analyzer) getFunctionBody(node ast.Node, lineNumber int, pass *analysis.Pass) (body *ast.BlockStmt, pos gotoken.Pos) {
+	ast.Inspect(node, func(n ast.Node) bool {
+		// Check if the node is a function declaration
+		funcDecl, ok := n.(*ast.FuncDecl)
+		if !ok {
+			return true
+		}
+
+		if pass.Fset.Position(n.Pos()).Line != lineNumber {
+			return true
+		}
+
+		body = funcDecl.Body
+		pos = n.Pos()
+		return false
+	})
+
+	return body, pos
+
+}
+
+func (p *Analyzer) checkGoroutineDefinition(ctx context.Context, pos gotoken.Position, pass *analysis.Pass) (gotoken.Pos, error) {
+	defFilePath, defLineNumber, err := p.lsp.Definition(ctx, pos.Filename, pos.Line, pos.Column)
+	if err != nil {
+		p.logger.Error("failed to find function definition", zap.Error(err))
+		return 0, err
+	}
+
+	file, err := p.parseFile(defFilePath, pass)
+	if err != nil {
+		p.logger.Error("failed to parse file", zap.Error(err))
+		return 0, err
+	}
+
+	body, defPosition := p.getFunctionBody(file, defLineNumber, pass)
+	return defPosition, p.checkGoroutine(body)
+}
+
+func (p *Analyzer) logLinterError(pass *analysis.Pass, errPos gotoken.Pos, callPos gotoken.Pos, err error) {
```

Comment:
Is this piece of code responsile for the final return value from the lint-panics utility? I was struggling to find where and what lint-panics returns. If so, then I suggest to:

```suggestion
func (p *Analyzer) reportLinterError(pass *analysis.Pass, errPos gotoken.Pos, callPos gotoken.Pos, err error) {
```

Reply:
Yes, this is the place of reporting, I will rename.
Reporting is done with `pass.Report(...)`. This is the standard way in `go/analysis` package to report issues. 

We launch the analyzer here:
https://github.com/status-im/status-go/blob/f718e5d645f591f62bc4a25cef434664e2b1688c/cmd/lint-panics/main.go#L34

And in the end it automatically prints all reported issues.